core/src/main/java/io/questdb/griffin/engine/functions/window/RankFunctionFactory.java

```diff
@@ -173,195 +249,267 @@ public void setColumnIndex(int columnIndex) {
 
-        private final static int VAL_CURRENT_INDEX = 1;
-        private final static int VAL_MAX_INDEX = 0;
-        private final static int VAL_OFFSET = 2;
-        private final Map map;
-        private final VirtualRecord partitionByRecord;
-        private final RecordSink partitionBySink;
-        private int columnIndex;
-        private RecordComparator recordComparator;
+        @Override
+        public void reopen() {
+            if (map != null) {
+                map.reopen();
+            }
+            rank = 0;
+        }
 
-        private long value;
+        @Override
+        public void reset() {
+            Misc.free(map);
+            rank = 0;
+        }
 
-        public RankFunction(Map map, VirtualRecord partitionByRecord, RecordSink partitionBySink) {
-            this.partitionByRecord = partitionByRecord;
-            this.partitionBySink = partitionBySink;
-            this.map = map;
+        @Override
+        public void toTop() {
+            super.toTop();
+            Misc.clear(map);
+            rank = 0;
         }
 
         @Override
         public void close() {
+            super.close();
             Misc.free(map);
             Misc.freeObjList(partitionByRecord.getFunctions());
         }
+    }
 
-        @Override
-        public void computeNext(Record record) {
-            partitionByRecord.of(record);
+    protected static class RankFunction extends LongFunction implements ScalarFunction, WindowFunction, Reopenable {
 
-            MapKey mapKey = map.withKey();
-            mapKey.put(partitionByRecord, partitionBySink);
-            MapValue mapValue = mapKey.createValue();
-            long maxIndex = 0;
-            if (mapValue.isNew()) {
-                mapValue.putLong(VAL_MAX_INDEX, 0);
-                mapValue.putLong(VAL_CURRENT_INDEX, 0);
-                mapValue.putLong(VAL_OFFSET, 0);
-            } else {
-                maxIndex = mapValue.getLong(VAL_MAX_INDEX);
-            }
+        private final CairoConfiguration configuration;
+        private final boolean dense;
+        private final String name;
+
+        private long rank;
+        private long count = 1;
+        private RecordComparator recordComparator;
+        private int columnIndex;
+        private SingleRecordSink singleRecordSinkA;
+        private SingleRecordSink singleRecordSinkB;
+        private long lastRecordOffset;
+        private RecordSink recordSink;
 
-            assert recordComparator == null;
-            value = maxIndex + 1;
-            mapValue.putLong(VAL_MAX_INDEX, value);
+        public RankFunction(CairoConfiguration configuration, boolean dense, String name) {
+            this.configuration = configuration;
+            this.dense = dense;
+            this.name = name;
         }
 
         @Override
         public long getLong(Record rec) {
-            assert recordComparator == null;
-            return value;
+            return rank;
```

Comment:
What happens when this function is called on an empty result set? We only reset `count` in `reopen`, `reset` and `toTop`, so `rank` field's value will stay invalid. Is that expected?